yahya/add-prompts-section-in-sdk (#31934)
DataDog/documentation: content/en/llm_observability/instrumentation/sdk.md

{{% /tab %}}
{{< /tabs >}}

## Prompt tracking

Attach structured prompt metadata to the LLM span so you can reproduce results, audit changes, and compare prompt performance across versions. When using templates, LLM Observability also provides [version tracking](#version-tracking) based on template content changes.

### Inline annotation (Python)

> Use `LLMObs.annotation_context(prompt=...)` to attach prompt metadata immediately before the LLM call. Use this for one-off or ad‑hoc prompts inside a function. The metadata is added to the current active span (for example, a span from a decorator or an auto-instrumented LLM call). For more details on span annotation, see [Annotating a span](#annotating-a-span).

Reviewer: I'd go straight to providing an example here rather than describing the usage. This is too many words for a dev which they're not going to read and/or too many words for them to get wrong

> Use `LLMObs.annotation_context(prompt=...)` to attach prompt metadata immediately before the LLM call. Use this for one-off or ad‑hoc prompts inside a function. The metadata is added to the current active span (for example, a span from a decorator or an auto-instrumented LLM call). For more details on span annotation, see [Annotating a span](#annotating-a-span).

Reviewer 2: it seems like all the text in the section only references `LLMObs.annotation_context` but all the examples in the section only use `LLMObs.annotate`. is that strange, or do users more or less understand the relationship between these by the time they are here

`prompt`
: required - dictionary
<br />A typed dictionary that follows the Prompt schema below.

{{% collapse-content title="Prompt structure" level="h4" expanded=false id="prompt-structure" %}}

Supported keys:

> - `id` (string): Logical identifier for this prompt. Should be unique per `ml_app`. Defaults to `{ml_app}-unnamed_prompt`

Reviewer: ```suggestion
- `id` (str): Logical identifier for this prompt. Should be unique per `ml_app`. Defaults to `{ml_app}-unnamed_prompt`
```

> - `version` (string): Version tag for the prompt (for example, "1.0.0"). See [version tracking](#version-tracking) for more details.

Reviewer: ```suggestion
- `version` (str): Version tag for the prompt (for example, "1.0.0"). See [version tracking](#version-tracking) for more details.
```

> - `variables` (object: Dict[str, str]): Variables used to render the template.

Reviewer: ```suggestion
- `variables` (Dict[str, str]): Variables used to render the template.
```

> - `template` (string): Single string template form.

Reviewer: ```suggestion
- `template` (str): Single string template form.
```

- `chat_template` (list of objects or Messages): Multi-message template form. Provide a list of `{ "role": "<role>", "template": "<template>" }` objects, or a list of `Message` objects.
- `tags` (object: Dict[str, str]): Tags to attach to the prompt run.

> - `rag_context_variables` (list of strings): Variable keys that contain ground-truth/context content. Used for [hallucination detection](/llm_observability/evaluations/managed_evaluations/?tab=openai#hallucination).

Reviewer: ```suggestion
- `rag_context_variables` (List[str]): Variable keys that contain ground-truth/context content. Used for [hallucination detection](/llm_observability/evaluations/managed_evaluations/?tab=openai#hallucination).
```

Reviewer: can we standardize all the types here to be direct Python types?

- `rag_query_variables` (list of strings): Variable keys that contain the user query. Used for [hallucination detection](/llm_observability/evaluations/managed_evaluations/?tab=openai#hallucination).

{{% /collapse-content %}}

Notes:
- Annotating a prompt is only available on LLM spans.

> - Place the annotation immediately before the provider call so it applies to the correct LLM span.

Reviewer: these two bullet points are crucial and why we should lead with an example here bc this is too easy to get wrong IMO

{{% /tab %}}
{{< /tabs >}}

## Prompt tracking

> Attach structured prompt metadata to the LLM span so you can reproduce results, audit changes, and compare prompt performance across versions. When using templates, LLM Observability also provides [version tracking](#version-tracking) based on template content changes.

Reviewer 3: ```suggestion
Attach structured prompt metadata to LLM spans so you can reproduce results, audit changes, and compare prompt performance across versions. When using templates, LLM Observability also provides [version tracking](#version-tracking) based on template content changes.
```

{{% /tab %}}
{{< /tabs >}}

## Prompt tracking

> 

Reviewer 3: ```suggestion
{{< callout url="#" btn_hidden="true" >}}
  Prompt Tracking for LLM Observability is in Preview. For access, <a href="/help">contact Datadog Support</a>.
{{< /callout >}}
```
